File: frontend/shader/contract/ts/ton.ts
// ============================================================================
// TON (The Open Network) Configuration
// ============================================================================

export interface TONConfig {
  rpcUrl: string;
  network: "mainnet" | "testnet";
}

// ============================================================================
// TON API Response Types
// ============================================================================

export interface TONAccountResponse {
  balance?: string;
  state?: string;
}

export interface TONTransactionResponse {
  transaction_id?: {
    hash?: string;
    lt?: string;
  };
  hash?: string;
}

/**
 * TON Connector
 * Connects to The Open Network (TON) blockchain
 */
export class TONConnector {
  private readonly config: TONConfig;

  constructor(config: TONConfig) {
    this.config = config;
  }

  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid TON address");
    }

    try {
      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = this.config.rpcUrl || "https://toncenter.com/api/v2";
      } else {
        apiUrl = this.config.rpcUrl || "https://testnet.toncenter.com/api/v2";
      }

      // TON addresses need to be converted to raw format for API calls
      // For simplicity, we'll use the address directly (in production, use @ton/core to convert)
      const response = await fetch(
        `${apiUrl}/getAddressInformation?address=${encodeURIComponent(address)}`,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to fetch TON balance");
      }

      const data = (await response.json()) as {
        result?: TONAccountResponse;
        ok?: boolean;
      };

      if (!data.ok || !data.result) {
        throw new Error("Failed to fetch TON account information");
      }

      // Return balance in nanograms (smallest unit, 1 TON = 1,000,000,000 nanograms)
      return data.result.balance || "0";
    } catch (error) {
      throw new Error(
        `TON getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid TON address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // TON transactions require:
      // 1. Creating transaction with proper message structure
      // 2. Signing with private key using TVM (TON Virtual Machine)
      // 3. Submitting to network
      // This is simplified - in production use @ton/core or tonweb

      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = this.config.rpcUrl || "https://toncenter.com/api/v2";
      } else {
        apiUrl = this.config.rpcUrl || "https://testnet.toncenter.com/api/v2";
      }

      // Get account state to get sequence number
      const accountResponse = await fetch(
        `${apiUrl}/getAddressInformation?address=${encodeURIComponent(
          fromAddress
        )}`,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      if (!accountResponse.ok) {
        throw new Error("Failed to get account information");
      }

      const accountData = (await accountResponse.json()) as {
        result?: TONAccountResponse;
        ok?: boolean;
      };

      if (!accountData.ok) {
        throw new Error("Failed to get account state");
      }

      // Send transaction
      // Note: In production, use @ton/core to properly build and sign transactions
      // Transaction would include:
      // - source: fromAddress
      // - destination: toAddress
      // - amount: amountNum (in nanograms)
      // - seqno: account sequence number
      // - body: message body (empty for simple transfers)
      const sendResponse = await fetch(`${apiUrl}/sendBoc`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          boc: "base64_encoded_signed_transaction", // In production, this would be the signed transaction BOC
        }),
      });

      if (!sendResponse.ok) {
        throw new Error("Failed to send TON transaction");
      }

      const sendData = (await sendResponse.json()) as {
        result?: TONTransactionResponse;
        ok?: boolean;
      };

      if (!sendData.ok || !sendData.result) {
        throw new Error("Transaction submission failed");
      }

      return (
        sendData.result.transaction_id?.hash ||
        sendData.result.hash ||
        "transaction_hash"
      );
    } catch (error) {
      throw new Error(
        `TON sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  isValidAddress(address: string): boolean {
    // TON addresses can be in different formats:
    // - Raw format: base64 encoded (48 chars)
    // - User-friendly format: starts with 0: or -1: followed by hex
    // - Bounceable/non-bounceable formats
    // Basic validation - check for common TON address patterns
    const rawPattern = /^[A-Za-z0-9+/]{48}$/; // Base64 raw format
    const friendlyPattern = /^(0|-1):[A-Fa-f0-9]{64}$/; // User-friendly format
    const bounceablePattern = /^[A-Za-z0-9_-]{48}$/; // Bounceable format

    return (
      rawPattern.test(address) ||
      friendlyPattern.test(address) ||
      bounceablePattern.test(address) ||
      address.length === 48
    );
  }

  /**
   * Bridge tokens from TON to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "TON";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["ton", "the-open-network", "telegram-open-network"];
  }
}

File: store/rust/rust/ton.rs
//! TON (The Open Network) connector

use crate::{BridgeError, BridgeResult, TONNetwork, http_get_json, http_post_json};
use serde::{Deserialize, Serialize};

/// TON connector
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TONConnector {
    rpc_url: String,
    #[allow(dead_code)]
    network: TONNetwork,
}

impl TONConnector {
    pub fn new(rpc_url: &str, network: TONNetwork) -> BridgeResult<Self> {
        Ok(Self { rpc_url: rpc_url.to_string(), network })
    }

    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // TON addresses need to be converted to raw format for API calls
        // For simplicity, we'll use the address directly (in production, use @ton/core to convert)
        // URL encode the address manually
        let encoded_address: String = address
            .chars()
            .map(|c| {
                if c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.' || c == '~' {
                    c.to_string()
                } else {
                    format!("%{:02X}", c as u8)
                }
            })
            .collect();
        let url = format!(
            "{}/getAddressInformation?address={}",
            self.rpc_url.trim_end_matches('/'),
            encoded_address
        );

        let result = http_get_json(&url)?;

        // Check if response is ok
        if let Some(ok) = result.get("ok").and_then(|o| o.as_bool()) {
            if !ok {
                return Err(BridgeError::NetworkError("Failed to fetch TON account information".to_string()));
            }
        }

        // Return balance in nanograms (smallest unit, 1 TON = 1,000,000,000 nanograms)
        let balance = result
            .get("result")
            .and_then(|r| r.get("balance"))
            .and_then(|b| b.as_str())
            .and_then(|s| s.parse::<u64>().ok())
            .unwrap_or(0);

        Ok(balance)
    }

    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in nanograms
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) || !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress("Invalid TON address".to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // TON transactions require:
        // 1. Creating transaction with proper message structure
        // 2. Signing with private key using TVM (TON Virtual Machine)
        // 3. Submitting to network
        // This is simplified - in production use @ton/core or tonweb
        let url = format!("{}/sendBoc", self.rpc_url.trim_end_matches('/'));
        let body = serde_json::json!({
            "boc": "base64_encoded_signed_transaction" // In production, this would be the signed transaction BOC
        });

        let result = http_post_json(&url, body)?;

        // Check if response is ok
        if let Some(ok) = result.get("ok").and_then(|o| o.as_bool()) {
            if !ok {
                return Err(BridgeError::TransactionFailed("Transaction submission failed".to_string()));
            }
        }

        result
            .get("result")
            .and_then(|r| r.get("transaction_id"))
            .and_then(|ti| ti.get("hash"))
            .and_then(|h| h.as_str())
            .map(|s| s.to_string())
            .or_else(|| {
                result
                    .get("result")
                    .and_then(|r| r.get("hash"))
                    .and_then(|h| h.as_str())
                    .map(|s| s.to_string())
            })
            .ok_or_else(|| BridgeError::TransactionFailed("Failed to get transaction hash".to_string()))
    }

    pub fn is_valid_address(&self, address: &str) -> bool {
        // TON addresses can be in different formats:
        // - Raw format: base64 encoded (48 chars)
        // - User-friendly format: starts with 0: or -1: followed by hex
        // - Bounceable/non-bounceable formats
        // Basic validation - check for common TON address patterns
        let raw_pattern = |s: &str| {
            s.len() == 48
                && s.chars()
                    .all(|c| c.is_ascii_alphanumeric() || c == '+' || c == '/' || c == '=')
        };
        let friendly_pattern = |s: &str| {
            (s.starts_with("0:") || s.starts_with("-1:"))
                && s.len() == 66
                && s[2..].chars().all(|c| c.is_ascii_hexdigit())
        };
        let bounceable_pattern = |s: &str| {
            s.len() == 48
                && s.chars()
                    .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
        };

        raw_pattern(address) || friendly_pattern(address) || bounceable_pattern(address) || address.len() == 48
    }
}

